### PR TITLE
RavenDB-11352 Pull Replication

### PR DESCRIPTION
**Pull request by karmeli87**

### CODE REVIEW

File: src/Raven.Client/Documents/Commands/GetRemoteTaskTopologyCommand.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        public GetRemoteTaskTopologyCommand(string remoteDatabase, string databaseGroupId, string remoteTask)
+        {
+            _remoteDatabase = remoteDatabase ?? throw new ArgumentNullException(nameof(remoteDatabase));
+            _databaseGroupId = databaseGroupId ?? throw new ArgumentNullException(nameof(databaseGroupId));
+            _remoteTask = remoteTask ?? throw new ArgumentNullException(nameof(remoteTask));
+        }
+
+        public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, ServerNode node, out string url)
+        {
+            url = $"{node.Url}/info/remote-task/topology?database={_remoteDatabase}&remote-task={_remoteTask}&groupId={_databaseGroupId}";
```
Comment (ayende) [inline src/Raven.Client/Documents/Commands/GetRemoteTaskTopologyCommand.cs]: `Uri.EscapeDataString` for the parameters

File: src/Raven.Client/Documents/Commands/GetTcpInfoForRemoteTaskCommand.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+        public GetTcpInfoForRemoteTaskCommand(string remoteDatabase, string remoteTask)
+        {
+            _remoteDatabase = remoteDatabase ?? throw new ArgumentNullException(nameof(remoteDatabase));
+            _remoteTask = remoteTask ?? throw new ArgumentNullException(nameof(remoteTask));
+        }
+
+        public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, ServerNode node, out string url)
+        {
+            url = $"{node.Url}/info/remote-task/tcp?database={_remoteDatabase}&remote-task={_remoteTask}";
```
Comment (ayende) [inline src/Raven.Client/Documents/Commands/GetTcpInfoForRemoteTaskCommand.cs]: `Uri.EscapeDataString` for the parameters

File: src/Raven.Client/Documents/Operations/OngoingTasks/GetTaskInfoResult.cs [hunk 3]
```diff
@@ -103,6 +105,54 @@ public override DynamicJsonValue ToJson()
...
+        public OngoingTaskPullReplicationAsSink()
+        {
+            TaskType = OngoingTaskType.PullReplicationAsSink;
+        }
+
+        public string DestinationUrl { get; set; }
+        public string[] TopologyDiscoveryUrls { get; set; }
+        public string DestinationDatabase { get; set; }
+        public string MentorNode { get; set; }
+        public string ConnectionStringName { get; set; }
```
Comment (ayende) [inline src/Raven.Client/Documents/Operations/OngoingTasks/GetTaskInfoResult.cs:142]: Why do we have a connection string here? And how do we define the cert?

File: src/Raven.Client/Documents/Operations/Replication/PutPullReplicationDefinitionOperation.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Documents.Operations.OngoingTasks;
+using Raven.Client.Http;
+using Raven.Client.Json;
+using Raven.Client.Json.Converters;
+using Sparrow.Json;
+
+namespace Raven.Client.Documents.Operations.Replication
+{
+    public class PutPullReplicationDefinitionOperation : IMaintenanceOperation<ModifyOngoingTaskResult>
```
Comment (ayende) [inline src/Raven.Client/Documents/Operations/Replication/PutPullReplicationDefinitionOperation.cs]: Missing `AsHub`

File: src/Raven.Client/ServerWide/DatabaseRecord.cs [hunk 5]
```diff
@@ -168,6 +172,8 @@ public void EnsureTaskNameIsNotUsed(string taskName)
 
             if (ExternalReplications.Any(x => x.Name.Equals(taskName, StringComparison.OrdinalIgnoreCase)))
                 throw new InvalidOperationException($"Can't use task name '{taskName}', there is already an External Replications task with that name");
+            if (SinkPullReplications.Any(x => x.Name.Equals(taskName, StringComparison.OrdinalIgnoreCase)))
```
Comment (ayende) [inline src/Raven.Client/ServerWide/DatabaseRecord.cs:175]: Should also have this for `HubPullReplications `

File: src/Raven.Server/Commercial/LicenseManager.cs [hunk 6]
```diff
@@ -1320,6 +1320,19 @@ public void AssertCanAddExternalReplication()
             throw GenerateLicenseLimit(LimitType.ExternalReplication, details);
         }
 
+        public void AssertCanAddPullReplication()
+        {
+            //TODO: add this feature to the license
```
Comment (ayende) [inline src/Raven.Server/Commercial/LicenseManager.cs:1325]: @grisha-kotler

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 7]
```diff
@@ -43,35 +43,66 @@ public class IncomingReplicationHandler : IDisposable
         public event Action<IncomingReplicationHandler> DocumentsReceived;
         public event Action<LiveReplicationPulsesCollector.ReplicationPulse> HandleReplicationPulse;
 
+        public void ClearEvents()
+        {
+            if (Failed != null)
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: `Failed = null` would do the same much easier.

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 8]
```diff
@@ -43,35 +43,66 @@ public class IncomingReplicationHandler : IDisposable
         public event Action<IncomingReplicationHandler> DocumentsReceived;
         public event Action<LiveReplicationPulsesCollector.ReplicationPulse> HandleReplicationPulse;
 
+        public void ClearEvents()
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:46]: why is this required?

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 9]
```diff
@@ -43,35 +43,66 @@ public class IncomingReplicationHandler : IDisposable
...
+
         public long LastDocumentEtag;
         public long LastHeartbeatTicks;
 
         private readonly ConcurrentQueue<IncomingReplicationStatsAggregator> _lastReplicationStats = new ConcurrentQueue<IncomingReplicationStatsAggregator>();
 
         private IncomingReplicationStatsAggregator _lastStats;
 
-        public IncomingReplicationHandler(TcpConnectionOptions options,
+        public readonly string PullReplicationName;
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:60]: Probably should add this to the `FromToString`

File: src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs [hunk 10]
```diff
@@ -70,13 +71,17 @@ public class OutgoingReplicationHandler : IDisposable
...
+        public ReplicationNode Destination;
         private readonly bool _external;
 
         private readonly ConcurrentQueue<OutgoingReplicationStatsAggregator> _lastReplicationStats = new ConcurrentQueue<OutgoingReplicationStatsAggregator>();
         private OutgoingReplicationStatsAggregator _lastStats;
         private readonly TcpConnectionInfo _connectionInfo;
 
+        // In case this is an outgoing pull replication from the hub
+        // we need to associate this instance to the replication definition. 
+        public string PullReplicationDefinitionName;
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs:83]: Add this to the `FromToString`

File: src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs [hunk 11]
```diff
@@ -107,248 +112,159 @@ public OutgoingReplicationStatsAggregator GetLatestReplicationPerformance()
...
+                        var isPullReplication = SendPreliminaryData();
+                        if (isPullReplication)
                         {
-                            const string msg = "Got operation canceled notification while opening outgoing replication channel. " +
-                                               "Aborting and closing the channel.";
-                            if (_log.IsInfoEnabled)
-                                _log.Info(msg, e);
-                            AddReplicationPulse(ReplicationPulseDirection.OutgoingInitiateError, msg);
-                            throw;
+                            InitiatePullReplicationAsEdge(supportedFeatures);
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs]: `AsSink`

File: src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs [hunk 12]
```diff
@@ -107,248 +112,159 @@ public OutgoingReplicationStatsAggregator GetLatestReplicationPerformance()
...
+                _stream = wrapSsl.Result;
+                _interruptibleRead = new InterruptibleRead(_database.DocumentsStorage.ContextPool, _stream);
+
+                using (_buffer = JsonOperationContext.ManagedPinnedBuffer.LongLivedInstance())
+                {
+                    var supportedFeatures = NegotiateReplicationVersion(authorizationInfo);
+                    if (supportedFeatures.Replication.PullReplication)
+                    {
+                        var isPullReplication = SendPreliminaryData();
+                        if (isPullReplication)
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs]: We just initiated the connection, how can we be pull replication here?

File: src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs [hunk 13]
```diff
@@ -107,248 +112,159 @@ public OutgoingReplicationStatsAggregator GetLatestReplicationPerformance()
...
+            };
 
-                            _waitForChanges.Reset();
-                        }
-                    }
-                }
+            using (_parent._server.Server._tcpContextPool.AllocateOperationContext(out var ctx))
+            using (ctx.GetManagedBuffer(out _buffer))
+            {
+                _parent.RunPullReplicationAsEdge(tcpOptions, _buffer, Destination as PullReplicationAsSink);
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs]: `AsSink`

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 14]
```diff
@@ -160,7 +163,128 @@ public ReplicationLoader(DocumentDatabase database, ServerStore server)
...
+            var externalReplication = pullReplicationDefinition.ToExternalReplication(initialRequest, taskId);
+            var outgoingReplication = new OutgoingReplicationHandler(this, Database, externalReplication, external: true, initialRequest.Info)
+            {
+                PullReplicationDefinitionName = initialRequest.PullReplicationDefinitionName
+            };
+
+            outgoingReplication.Failed += OnOutgoingSendingFailed;
+            outgoingReplication.SuccessfulTwoWaysCommunication += OnOutgoingSendingSucceeded;
+            _outgoing.TryAdd(outgoingReplication); // can't fail, this is a brand new instance
+            Console.WriteLine($"({_server.GetNodeHttpServerUrl()}) Hub to {initialRequest.SourceUrl}");
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs]: Remove this

Comment (ayende) [reply src/Raven.Server/Documents/Replication/ReplicationLoader.cs]: Or change to log statement

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 15]
```diff
@@ -160,7 +163,128 @@ public ReplicationLoader(DocumentDatabase database, ServerStore server)
...
+
+            outgoingReplication.StartPullReplicationAsHub(tcpConnectionOptions.Stream, supportedVersions);
+            OutgoingReplicationAdded?.Invoke(outgoingReplication);
+        }
+
+        public void RunPullReplicationAsEdge(TcpConnectionOptions tcpConnectionOptions, JsonOperationContext.ManagedPinnedBuffer buffer, PullReplicationAsSink destination)
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs]: `AsSink`

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 16]
```diff
@@ -423,11 +532,67 @@ private void HandleTopologyChange(DatabaseRecord newRecord)
             DisposeConnections(instancesToDispose);
         }
 
-        private void DisposeConnections(List<OutgoingReplicationHandler> instancesToDispose)
+        private void HandleCentralPullReplication(DatabaseRecord newRecord, List<IDisposable> instancesToDispose)
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs]: `Hub`

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 17]
```diff
@@ -423,11 +532,67 @@ private void HandleTopologyChange(DatabaseRecord newRecord)
...
+            }
+        }
+
+        private void DropIncomingConnections(IEnumerable<ReplicationNode> connectionsToRemove, List<IDisposable> instancesToDispose)
+        {
+            var toRemove = connectionsToRemove?.ToList();
+            if (toRemove == null || toRemove.Count  == 0)
+                return;
+
+            // this is relevant for edge pull replication
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs]: sink

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 18]
```diff
@@ -423,11 +532,67 @@ private void HandleTopologyChange(DatabaseRecord newRecord)
...
+            foreach (var incoming in _incoming)
+            {
+                var instance = incoming.Value;
+                if (toRemove.Any(conn => conn.Url == instance.ConnectionInfo.SourceUrl))
+                {
+                    if (_incoming.TryRemove(instance.ConnectionInfo.SourceDatabaseId, out _))
+                        IncomingReplicationRemoved?.Invoke(instance);
+
+                    instance.ClearEvents();
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs:584]: Why do you need to call this?

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 19]
```diff
@@ -744,9 +950,89 @@ private TcpConnectionInfo GetConnectionInfo(ReplicationNode node, bool external)
...
+                    remoteDatabaseUrls = cmd.Result;
+                }
+
+                // fetch tcp info for the central nodes
+                using (var requestExecutor = RequestExecutor.CreateForFixedTopology(remoteDatabaseUrls,
+                    pullReplicationAsSink.ConnectionString.Database, certificate, DocumentConventions.Default))
+                {
+                    var cmd = new GetTcpInfoForRemoteTaskCommand(database, remoteTask);
+                    requestExecutor.Execute(cmd, ctx);
+                    pullReplicationAsSink.Database = database;
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs]: Why are you mutating the `pullReplicationAsSink` here?

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 20]
```diff
@@ -744,9 +950,89 @@ private TcpConnectionInfo GetConnectionInfo(ReplicationNode node, bool external)
...
+
+            if (edge.CertificateWithPrivateKey == null)
+                return _server.Server.Certificate.Certificate;
+
+            var certBytes = Convert.FromBase64String(edge.CertificateWithPrivateKey);
+            return new X509Certificate2(certBytes, edge.CertificatePassword, X509KeyStorageFlags.Exportable | X509KeyStorageFlags.MachineKeySet);
+
+        }
+
+        public (string Url, OngoingTaskConnectionStatus Status) GetPullReplicationAsEdgeDestination(string name)
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs]: `AsSink`

File: src/Raven.Server/RavenServer.cs [hunk 21]
```diff
@@ -1594,6 +1614,22 @@ private bool TryAuthorize(RavenConfiguration configuration, Stream stream, TcpCo
...
+                    if (info != null && info.AuthorizeAs == TcpConnectionHeaderMessage.AuthorizationInfo.AuthorizeMethod.PullReplication)
+                    {
+                        using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext ctx))
+                        using (ctx.OpenReadTransaction())
+                        {
+                            if (ServerStore.Cluster.TryReadPullReplicationDefinition(header.DatabaseName, info.AuthorizationFor, ctx, out var pullReplication)
+                                && pullReplication.CanAccess(certificate.Thumbprint))
+                                return true;
+
+                            msg = "The certificate " + certificate.FriendlyName + " does not allow access to " + header.DatabaseName;
```
Comment (ayende) [inline src/Raven.Server/RavenServer.cs]: Also include ` info.AuthorizationFor` in the error

File: src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs [hunk 22]
```diff
@@ -19,5 +31,153 @@ public Task Get()
...
+                        }
+                        throw new AuthorizationException(PullReplicationAuthorizationExceptionMessage(database, remoteTask));
+                    }
+
+                default:
+                    throw new ArgumentException($"This is a bug, we should deal with '{feature?.Status}' authentication status at RequestRoute.TryAuthorize function.");
+            }
+        }
+
+        private static string PullReplicationAuthorizationExceptionMessage(string database, string remoteTask)
```
Comment (ayende) [inline src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs]: Let's use the same error message on the TCP side as well

File: src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs [hunk 23]
```diff
@@ -19,5 +31,153 @@ public Task Get()
...
+        {
+            return $"Cannot connect to '{remoteTask}' on '{database}'. The database or task may not exists or you don't have the credentials to access it.";
+        }
+
+        private List<string> GetResponsibleNodes(DatabaseTopology topology, string databaseGroupId, PullReplicationDefinition pullReplication)
+        {
+            var list = new List<string>();
+            // we distribute connections to have load balancing when many edges are connected.
```
Comment (ayende) [inline src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs]: sinks

File: src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs [hunk 24]
```diff
@@ -19,5 +31,153 @@ public Task Get()
...
+            // this is the central cluster, so we make the decision which node will do the pull replication only once and only here,
```
Comment (ayende) [inline src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs]: hub

File: src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs [hunk 25]
```diff
@@ -19,5 +31,153 @@ public Task Get()
...
+            // for that we create a dummy IDatabaseTask.
+            var mentorNodeTask = new PullNodeTask
+            {
+                Mentor = pullReplication.MentorNode,
+                DatabaseGroupId = databaseGroupId
+            };
+
+            while (topology.Members.Count > 0)
+            {
+                var next = topology.WhoseTaskIsIt(ServerStore.CurrentRachisState, mentorNodeTask, null);
```
Comment (ayende) [inline src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs:149]: If `Mentor` is defined, can we get it here multiple times?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Documents/Operations/OngoingTasks/GetTaskInfoResult.cs (hunk 3 above)
Comment (karmeli87) [reply src/Raven.Client/Documents/Operations/OngoingTasks/GetTaskInfoResult.cs:142]: The connection string contains the entry points to the hub cluster.

This class is only to view the current status of the ongoing task. 
The certificate is defined with the entire Sink at the `/databases/*/admin/tasks/sink-pull-replication` endpoint

File: src/Raven.Client/ServerWide/DatabaseRecord.cs (hunk 5 above)
Comment (karmeli87) [reply src/Raven.Client/ServerWide/DatabaseRecord.cs:175]: `HubPullReplications` is dictionary so we don't need to worry about duplicate names.

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 8 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:46]: Usually we don't care about this because we unregistered when an exception occurred.
Here we gracefully closing the Sink.

File: src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs (hunk 25 above)
Comment (karmeli87) [reply src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs:149]: No, we remove the next node from the member list.

File: src/Raven.Client/Documents/Operations/OngoingTasks/GetTaskInfoResult.cs (hunk 3 above)
Comment (ayende) [reply src/Raven.Client/Documents/Operations/OngoingTasks/GetTaskInfoResult.cs:142]: I don't see any other references to connection strings in pull replication. Where is this name from?

File: src/Raven.Client/ServerWide/DatabaseRecord.cs (hunk 5 above)
Comment (ayende) [reply src/Raven.Client/ServerWide/DatabaseRecord.cs:175]: What you are doing is making sure that there aren't duplicated names across different types of stuff.
sink pull, external repl, etc.
It isn't for correctness, it is to avoid confusion with the user

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 8 above)
Comment (ayende) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:46]: Just abandoning this instance should work well enough, no?

Comment (karmeli87) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:46]: Since we dispose in an `async` manner I worry about race between disposing and entering those events.

File: src/Raven.Client/Documents/Operations/OngoingTasks/GetTaskInfoResult.cs (hunk 3 above)
Comment (karmeli87) [reply src/Raven.Client/Documents/Operations/OngoingTasks/GetTaskInfoResult.cs:142]: `PullReplicationAsSink` inherits from `ExternalReplication` and treated as such on the very beginning.
